akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,325 @@
+      val effect = setup.commandHandler(commandContext, state.state, cmd)
+      applyEffects(cmd, state, effect.asInstanceOf[EffectImpl[E, S]]) // TODO can we avoid the cast?
+    }
+
+    @tailrec def applyEffects(
+      msg:         Any,
+      state:       EventsourcedState[S],
+      effect:      EffectImpl[E, S],
+      sideEffects: immutable.Seq[ChainableEffect[_, S]] = Nil
+    ): Behavior[InternalProtocol] = {
+      if (log.isDebugEnabled)
+        log.debug(s"Handled command [{}], resulting effect: [{}], side effects: [{}]", msg.getClass.getName, effect, sideEffects.size)
+
+      effect match {
+        case CompositeEffect(eff, currentSideEffects) ⇒
+          // unwrap and accumulate effects
+          applyEffects(msg, state, eff, currentSideEffects ++ sideEffects)
+
+        case Persist(event) ⇒
+          // apply the event before persist so that validation exception is handled before persisting
+          // the invalid event, in case such validation is implemented in the event handler.
+          // also, ensure that there is an event handler for each single event
+          val newState = state.applyEvent(setup, event)
+          val eventToPersist = tagEvent(event)
+
+          val newState2 = internalPersist(newState, eventToPersist)
+
+          val handler: Any ⇒ Unit = { _ ⇒
+            if (setup.snapshotWhen(newState2.state, event, newState2.seqNr))
+              internalSaveSnapshot(newState2)
+          }
+          val pendingInvocations = StashingHandlerInvocation(event, handler) :: Nil
+
+          persistingEvents(newState2, pendingInvocations, sideEffects)
+
+        case PersistAll(events) ⇒
+          if (events.nonEmpty) {
+            // apply the event before persist so that validation exception is handled before persisting
+            // the invalid event, in case such validation is implemented in the event handler.
+            // also, ensure that there is an event handler for each single event
+            var count = events.size
+            var seqNr = state.seqNr
+            val (newState, shouldSnapshotAfterPersist) = events.foldLeft((state, false)) {
+              case ((currentState, snapshot), event) ⇒
+                seqNr += 1
+                val shouldSnapshot = snapshot || setup.snapshotWhen(currentState.state, event, seqNr)
+                (currentState.applyEvent(setup, event), shouldSnapshot)
+            }
+
+            val eventsToPersist = events.map(tagEvent)
+
+            val newState2 = internalPersistAll(eventsToPersist, newState)
+
+            val handler: Any ⇒ Unit = { _ ⇒
+              count -= 1
+              if (count == 0) {
+                if (shouldSnapshotAfterPersist)
+                  internalSaveSnapshot(newState2)
+              }
+            }
+
+            val pendingInvocations = events.map { event ⇒
+              StashingHandlerInvocation(event, handler)
+            }
```

Comment:
ok, here we have more than one callback, but we don't really need this pendingInvocations stack here. We can have an event counter and do the stuff when that many `WriteMessageSuccess` have been received.
We actually don't need the callback at all, since we only use it for snapshotting.
Meaning that we can replace `pendingInvocations` with `numberOfEvents: Int` and `snapshotAfterPersist: Boolean`.

Let me know if these comments are difficult to follow, and I can do that refactoring instead.